src/Confluent.Kafka/Consumer.cs

```diff
@@ -177,37 +179,85 @@ private void LogCallback(IntPtr rk, SyslogLevel level, string fac, string buf)
                     throw new Exception("Unexpected rebalance callback on disposed kafkaHandle");
                 }
 
+                if (kafkaHandle.RebalanceProtocol == "COOPERATIVE" &&
+                    this.revokedOrLostHandlerIsFunc)
+                {
+                    throw new InvalidOperationException("Neither revoked nor lost partition handlers may return an updated assignment when a COOPERATIVE assignor is in use");
+                }
+
+                var partitions = SafeKafkaHandle.GetTopicPartitionOffsetErrorList(partitionsPtr).Select(p => p.TopicPartition).ToList();
+
                 if (err == ErrorCode.Local_AssignPartitions)
                 {
                     if (partitionsAssignedHandler == null)
                     {
-                        Assign(partitionAssignment.Select(p => new TopicPartitionOffset(p, Offset.Unset)));
+                        if (kafkaHandle.RebalanceProtocol == "COOPERATIVE")
+                        {
+                            IncrementalAssign(partitions.Select(p => new TopicPartitionOffset(p, Offset.Unset)));
+                        }
+                        else
+                        {
+                            Assign(partitions.Select(p => new TopicPartitionOffset(p, Offset.Unset)));
+                        }
                         return;
                     }
 
                     lock (assignCallCountLockObj) { assignCallCount = 0; }
-                    var assignTo = partitionsAssignedHandler(partitionAssignment);
+                    var assignTo = partitionsAssignedHandler(partitions);
                     lock (assignCallCountLockObj)
                     {
                         if (assignCallCount > 0)
                         {
-                            throw new InvalidOperationException("Assign/Unassign must not be called in the partitions assigned handler.");
+                            throw new InvalidOperationException("(Incremental)Assign/Unassign must not be called in the partitions assigned handler");
                         }
                     }
-                    Assign(assignTo);
+
+                    if (kafkaHandle.RebalanceProtocol == "COOPERATIVE")
+                    {
+                        if (assignTo.Count() != partitions.Count())
+                        {
+                            throw new InvalidOperationException("The partitions assigned handler must not return a different set of topic partitions than it was provided");
```

Comment:
Do we really have to enforce this?
Some applications want to consume the subscribed topics and a common command topic/partition that all consumers read.

Reply:
I think so. The application logic in the cooperative case will get pretty complex, i think we just don't want to go there.

This API we have for realizing command topic/partition IMO is bad:

1. structurally, it should be represented to the user as a separate thing - the two things shouldn't be as interleaved as they are. it is not the clearest way of modelling the requirements.
2. the terminology is ambiguous / tricky and it makes the API harder to understand.
3. Java client does not allow this.
4. It makes the CGRP implementation harder! 

I think best practice currently for people who want this pattern is to just use two consumers. Ideally we'd be able to use a single instance for this (but I think two consumers is not a big deal for most users).